tests_integration/test_intro.py

```diff
@@ -0,0 +1,273 @@
+        ]
+
+        def validate_context_and_answer_and_hops(example, pred, trace=None):
+            if not dspy.evaluate.answer_exact_match(example, pred):
+                return False
+            if not dspy.evaluate.answer_passage_match(example, pred):
+                return False
+
+            hops = [example.question] + [outputs.query for *_, outputs in trace if "query" in outputs]
+
+            if max([len(h) for h in hops]) > 100:
+                return False
+            if any(
+                dspy.evaluate.answer_exact_match_str(hops[idx], hops[:idx], frac=0.8) for idx in range(2, len(hops))
+            ):
+                return False
+
+            return True
+
+        teleprompter = BootstrapFewShot(metric=validate_context_and_answer_and_hops)
+        compiled_baleen = teleprompter.compile(
+            SimplifiedBaleen(),
+            teacher=SimplifiedBaleen(passages_per_hop=2),
+            trainset=trainset,
+        )
+        uncompiled_baleen_retrieval_score = evaluate_on_hotpotqa(uncompiled_baleen, metric=gold_passages_retrieved)
+        compiled_baleen_retrieval_score = evaluate_on_hotpotqa(compiled_baleen, metric=gold_passages_retrieved)
+
+        # Assertions for the retrieval scores
+        assert f"## Retrieval Score for RAG: {compiled_rag_retrieval_score}" == "## Retrieval Score for RAG: 26.0"
+        assert (
+            f"## Retrieval Score for uncompiled Baleen: {uncompiled_baleen_retrieval_score}"
+            == "## Retrieval Score for uncompiled Baleen: 36.0"
+        )
+        assert (
+            f"## Retrieval Score for compiled Baleen: {compiled_baleen_retrieval_score}"
+            == "## Retrieval Score for compiled Baleen: 60.0"
+        )
+        assert compiled_baleen("How many storeys are in the castle that David Gregory inherited?") is not None
+
+    def assert_retrieval(self, dev_example) -> None:
+        retrieve = dspy.Retrieve(k=3)
+        top_k_passages = retrieve(dev_example.question).passages
+
+        # Assertions to verify the retrieval functionality
+        assert retrieve.k == 3
+        assert (
+            dev_example.question
+            == "What is the nationality of the chef and restaurateur featured in Restaurant: Impossible?"
+        )
+        assert (
+            top_k_passages[0]
+            == "Restaurant: Impossible | Restaurant: Impossible is an American reality television series, featuring "
+            "chef and restaurateur Robert Irvine, that aired on Food Network from 2011 to 2016."
+        )
+        assert (
+            top_k_passages[1]
+            == "Jean Joho | Jean Joho is a French-American chef and restaurateur. He is chef/proprietor of Everest in "
+            "Chicago (founded in 1986), Paris Club Bistro & Bar and Studio Paris in Chicago, The Eiffel Tower "
+            "Restaurant in Las Vegas, and Brasserie JO in Boston."
+        )
+        assert top_k_passages[2][:30] == "List of Restaurant: Impossible"
```

Comment:
small nitpick but any reason not to check the full passage and retrieved value below? :)

Reply:
anyways changed it to full paragraph 🗡️